### PR TITLE
Implement Connection Pooling and Threading

### PR DESCRIPTION
**Pull request by MrLight**
This implements Connection Pooling for the Snowflake Connection.
The code today doesn't reuse the snowflake connection. After each query the connection is dropped. This PullRequest will add ConnectionPooling to store open connections for later reuse. The poolsize and the connection timeout can be configured in the PluginConfig.
When the snowflake warehouse is in hard load it could be possible, that too much queries will be forwarded to snowflake, which will end in an overloaded Snowflake Warehouse. The poolsize will limit the concurrent queries which are forwarded to snowflake, but all additional Grafana queries will be stored in the query queue of the plugin. To drop new queries in the situation of overload a max. queued queries option is available. If more queries as configured are waiting a error will be returned to the panel. This will give snowflake the possibility to recover.
To implement the options in the ConfigEditor the editor code is changed from the legacy forms to the newer grafana forms.

If a Panel has multiple queries, the snowflake query will be done in multiple threads after this change.

All changes are speeding up the query execution and will reduce snowflake cloud query costs for connection creation.

This is the first pull-request. A second one implementing plugin-side caching is prepared. Please let me know if you are willing to implement such huge changes into your code base.

### CODE REVIEW

**Comment by devnied** [comment]
Thanks, for this MR, I will review it asap.
For your second MR "implementing plugin-side caching" I don't think this should be a functionality of this plugin as Snowflake already perform cache query result, without extra cost.

**Comment by MrLight** [comment]
Thanks, for looking into this.
Regarding the plugin-side caching. I have a running test version of that. I do some cleaning at the moment and will publish that in my fork first. Possibly you like to take a look into it. 
Out of my opinion it really helps in terms of usability, especially if you are browsing back n force or if someone sets (too) high refresh rates ...
Will ping you if its ready... (btw. thanks for merging my other pull request)

**Comment by MrLight** [comment]
I have published a beta release in "my" fork. If you like you can have look there...
[preview version](https://github.com/MrLight/snowflake-grafana-datasource/releases/tag/v1.8.0-beta)

File: pkg/snowflake_test.go [hunk 1]
```diff
@@ -16,8 +17,8 @@ func TestGetConfig(t *testing.T) {
 		response string
 		err      string
 	}{
-		{json: "{}", config: pluginConfig{}},
-		{json: "{\"account\":\"test\"}", config: pluginConfig{Account: "test"}},
+		{json: "{}", config: pluginConfig{ConnectionLifetime: "60", IntConnectionLifetime: 60, MaxOpenConnections: "100", IntMaxOpenConnections: 100, MaxQueuedQueries: "400", IntMaxQueuedQueries: 400}},
+		{json: "{\"account\":\"test\"}", config: pluginConfig{Account: "test", ConnectionLifetime: "60", IntConnectionLifetime: 60, MaxOpenConnections: "100", IntMaxOpenConnections: 100, MaxQueuedQueries: "400", IntMaxQueuedQueries: 400}},
```
Comment (devnied) [inline pkg/snowflake_test.go:21]: Can you please add an unit tests for custom configuration for parmater ConnectionLifetime, MaxOpenConnections,MaxQueuedQueries with good and bad values.

**Review by devnied — COMMENTED** [review]
Thank you for this great PR! 
I have a few pieces of feedback based on my review. 
I also need to test a few edge cases before I can approve it.

File: pkg/query.go [hunk 2]
```diff
@@ -185,19 +214,39 @@ func (qc *queryConfigStruct) transformQueryResult(columnTypes []*sql.ColumnType,
...
+
 	var qm queryModel
 	err := json.Unmarshal(dataQuery.JSON, &qm)
 	if err != nil {
-		log.DefaultLogger.Error("Could not unmarshal query", "err", err)
-		response.Error = err
-		return response
+		//log.DefaultLogger.Error("Could not unmarshal query", "err", err)
+		//queryResult.dataResponse.Error = err
+		panic("Could not unmarshal query")
```
[the remark this answers is not in the document]
Comment (MrLight) [reply pkg/query.go:243]: Due to the multi-thread approach this is done in the defer func() see line# 224-233. I will remove the comment. The panic will only panic the thread. See as reference: [grafana example](https://github.com/grafana/grafana/blob/main/pkg/tsdb/grafana-postgresql-datasource/sqleng/sql_engine.go#L116)
The panic is only for the queries with missing QueryText. If it is available an error is returned in respect to the QueryText. See line# 274 / 261

File: pkg/query.go [hunk 3]
```diff
@@ -185,19 +214,39 @@ func (qc *queryConfigStruct) transformQueryResult(columnTypes []*sql.ColumnType,
...
+	defer func() {
+		if r := recover(); r != nil {
+			log.DefaultLogger.Error("ExecuteQuery panic", "error", r, "stack", string(debug.Stack()))
+			if theErr, ok := r.(error); ok {
+				queryResult.dataResponse.Error = theErr
+			} else if theErrString, ok := r.(string); ok {
+				queryResult.dataResponse.Error = fmt.Errorf(theErrString)
+			} else {
+				//queryResult.dataResponse.Error = fmt.Errorf("unexpected error - %s", td.userError)
+			}
```
[the remark this answers is not in the document]
Comment (MrLight) [reply pkg/query.go:233]: done

File: pkg/snowflake.go [hunk 4]
```diff
@@ -45,40 +52,85 @@ type SnowflakeDatasource struct {
...
-	password := req.PluginContext.DataSourceInstanceSettings.DecryptedSecureJSONData["password"]
+	/*password := req.PluginContext.DataSourceInstanceSettings.DecryptedSecureJSONData["password"]
 	privateKey := req.PluginContext.DataSourceInstanceSettings.DecryptedSecureJSONData["privateKey"]
 
 	config, err := getConfig(req.PluginContext.DataSourceInstanceSettings)
 	if err != nil {
 		log.DefaultLogger.Error("Could not get config for plugin", "err", err)
 		return response, err
+	}*/
+	i, err := td.im.Get(ctx, req.PluginContext)
```
[the remark this answers is not in the document]
Comment (MrLight) [reply pkg/snowflake.go:65]: done

File: pkg/snowflake.go [hunk 5]
```diff
@@ -45,40 +52,85 @@ type SnowflakeDatasource struct {
...
+		return nil, err
+	}
+	instance := i.(*instanceSettings)
+	ch := make(chan DBDataResponse, len(req.Queries))
+	var wg sync.WaitGroup
+	// Execute each query in a goroutine and wait for them to finish afterwards
+	for _, query := range req.Queries {
+		wg.Add(1)
+		go td.query(ctx, &wg, ch, instance, query)
+		//go e.executeQuery(query, &wg, ctx, ch, queryjson)
```
[the remark this answers is not in the document]
Comment (MrLight) [reply pkg/snowflake.go:76]: done

**Comment by MrLight** [comment]
I have reworked the code, so that it should fit into the new structure. I can't test the oauth stuff, but should work as expected.

**Comment by devnied** [comment]
Thanks for the rebase.
Before merging, I need to check whether using the default Go connection pool settings is a better approach: ([DB.SetMaxOpenConns](https://pkg.go.dev/database/sql#DB.SetMaxOpenConns), [DB.SetMaxIdleConns](https://pkg.go.dev/database/sql#DB.SetMaxIdleConns), [DB.SetConnMaxIdleTime](https://pkg.go.dev/database/sql#DB.SetConnMaxIdleTime), [DB.SetConnMaxLifetime](https://pkg.go.dev/database/sql#DB.SetConnMaxLifetime)) https://go.dev/doc/database/manage-connections

**Comment by MrLight** [comment]
> Thanks for the rebase. Before merging, I need to check whether using the default Go connection pool settings is a better approach: ([DB.SetMaxOpenConns](https://pkg.go.dev/database/sql#DB.SetMaxOpenConns), [DB.SetMaxIdleConns](https://pkg.go.dev/database/sql#DB.SetMaxIdleConns), [DB.SetConnMaxIdleTime](https://pkg.go.dev/database/sql#DB.SetConnMaxIdleTime), [DB.SetConnMaxLifetime](https://pkg.go.dev/database/sql#DB.SetConnMaxLifetime)) https://go.dev/doc/database/manage-connections

Not really sure, what you are referencing to. The idea is exactly this. The Connection creation logic has moved into the `snowflake.go` file. 

```
        db, err := sql.Open("snowflake", connectionString)
	if err != nil {
		return nil, err
	}

	db.SetMaxOpenConns(int(config.IntMaxOpenConnections))
	db.SetMaxIdleConns(int(config.IntMaxOpenConnections))
	db.SetConnMaxLifetime(time.Duration(int(config.IntConnectionLifetime)) * time.Minute)
```

With this go-lang will create a connection pool with multiple connections. 
In my prod-setup I have a high amount of clients sending queries to snowflake. I had cases when the warehouse was overloaded and a lot of request where queued in snowflake. To avoid a deadlook in grafana a second counter ( called `queryCounter`) is implemented. If the querycounter limit is exceeded we will cancel the grafana query. 

```
        if int(qs.intMaxQueuedQueries) > 0 && int(qs.queryCounter.get()) >= (int(qs.intMaxQueuedQueries)) {
		err := errors.New("too many queries in queue. Check Snowflake connectivity or increase MaxQueuedQeries count")
		log.DefaultLogger.Error("Poolsize exceeded", "query", qs.qc.FinalQuery, "err", err)
		return result, err
	}
``` 

Concept is that we do not need to execute each and every query in case of Warehouse overload. We will drop all new queries until the warehouse is responsive again.
